module/util/log.go

```diff
@@ -8,68 +8,108 @@ import (
+// The function can be called concurrently. addProgress is the amount to add to the progress.
+// It is any integer number type, but all negative values are ignored.
+type LogProgressFunc[T int | uint | int32 | uint32 | uint64 | int64] func(addProgress T)
+
+type LogProgressConfig[T int | uint | int32 | uint32 | uint64 | int64] struct {
+	// message is part of the messages that will be logged.
+	// The full template is: `%s progress %d/%d (%.1f%%) total time %s`.
+	message string
+	// total is the total value of progress expected.
+	// When total is added to LogProgressFunc the progress is considered to be 100%.
+	total T
+	// noDataLogDuration. If the last log line was more than this duration ago and a new data point is added, a new log line is logged.
+	// No line is logged if no data is received. The minimum resolution for noDataLogDuration is 1 millisecond.
+	noDataLogDuration time.Duration
+	// ticks is the number of increments to log at. If total is > 0 there will be at least 2 ticks. One at 0 and one at total.
+	// If you want to log at every 10% set ticks to 11 (one is at 0%).
+	// If the number of ticks is more than total, it will be set to total + 1.
+	ticks uint64
 }
 
-func DefaultLogProgressConfig(
+// DefaultLogProgressConfig returns a LogProgressConfig with default values.
+// The default values will log every 10% and will log an additional line if new data is received
+// after no data has been received for 1 minute.
+func DefaultLogProgressConfig[T int | uint | int32 | uint32 | uint64 | int64](
 	message string,
-	total int,
-) LogProgressConfig {
-	nth := uint32(total / 10) // sample every 10% by default
-	if nth == 0 {
-		nth = 1
-	}
-
-	sampler := newProgressLogsSampler(nth, 60*time.Second)
-	return NewLogProgressConfig(
+	total T,
+) LogProgressConfig[T] {
+	return NewLogProgressConfig[T](
 		message,
 		total,
-		sampler,
+		60*time.Second,
+		10,
 	)
 }
 
-func NewLogProgressConfig(
+// NewLogProgressConfig creates and returns a new LogProgressConfig with the specified message, total, duration, and ticks.
+// The duration is rounded to the nearest millisecond.
+// The number of ticks is the number of increments to log at. Logging at 0% is always done.
+// If you want to log at 10% increments, set ticks to 10.
+func NewLogProgressConfig[T int | uint | int32 | uint32 | uint64 | int64](
 	message string,
-	total int,
-	sampler zerolog.Sampler) LogProgressConfig {
-	return LogProgressConfig{
-		Message: message,
-		Total:   total,
-		Sampler: sampler,
+	total T,
+	noDataLogDuration time.Duration,
+	ticks uint64,
+) LogProgressConfig[T] {
+	// sanitize total
+	if total < 0 {
+		total = 0
 	}
 
-}
+	// add the tick at 0%
+	ticks = ticks + 1
+
+	// sanitize ticks
+	// number of ticks should be at most total + 1
+	if uint64(total+1) < ticks {
+		ticks = uint64(total + 1)
+	}
+
+	return LogProgressConfig[T]{
+		message:           message,
+		total:             total,
+		noDataLogDuration: noDataLogDuration,
```

Comment:
Maybe also sanitize `noDataLogDuration` to make it use minimum of 1 millisecond, to make it consistent with code comment at:

https://github.com/onflow/flow-go/blob/7396328c81d292c3a5bb1ebd3865e25e8dd96ded/module/util/log.go#L23-L25